pkg/typesystem/reachabilitygraph.go

```diff
@@ -142,6 +144,78 @@ func (rg *ReachabilityGraph) RelationsEncounteredForResource(
 	return relationRefs, nil
 }
 
+// RelationsEncounteredForSubject returns all relations that are encountered when walking outward from a subject+relation.
+func (rg *ReachabilityGraph) RelationsEncounteredForSubject(
+	ctx context.Context,
+	allDefinitions []*core.NamespaceDefinition,
+	startingSubjectType *core.RelationReference,
+) ([]*core.RelationReference, error) {
+	if startingSubjectType.Namespace != rg.ts.nsDef.Name {
+		return nil, spiceerrors.MustBugf("gave mismatching namespace name for subject type to reachability graph")
+	}
+
+	allRelationNames := mapz.NewSet[string]()
+
+	subjectTypesToCheck := []*core.RelationReference{startingSubjectType}
+
+	// TODO(jschorr): optimize this to not require walking over all types recursively.
```

Comment:
We may want to add some sort of max depth to prevent infinite recursion that could happen via the schema? Perhaps adding a context deadline if this is taking too long as a precaution?

Reply:
This is a structural walk, not a a data-driven one; there should be no case where it causes infinite recursion as the check below reduces the working set over time